### PR TITLE
Add lock while writing to console and closing it

### PR DESCRIPTION
**Pull request by runcom**
This patch attemps to fix server side race found in #6148
With a `-race` build the race can be triggered just running:

```
docker run -ti busybox /bin/sh
Ctrl^d
[DATA RACE HERE AND DAEMON SIDE]
```

@LK4D4 I don't experience any lockup with the commands you posted in the issue so maybe something was already fixed about this (not sure, in case I'll close this)

Signed-off-by: Antonio Murdaca runcom@linux.com

### CODE REVIEW

**Comment by LK4D4** [comment]
I don't know :) maybe it's ok.

**Comment by runcom** [comment]
I also wonder if libcontainer also needs a lock in console_linux

**Comment by LK4D4** [comment]
@runcom I'm not a fan of locks in libcontainer :)

**Comment by runcom** [comment]
@LK4D4 yeah me neither, was just wondering waiting your comment :smile:

File: daemon/execdriver/native/driver.go [hunk 1]
```diff
@@ -409,7 +412,12 @@ func (t *TtyConsole) AttachPipes(pipes *execdriver.Pipes) error {
 }
 
 func (t *TtyConsole) Close() error {
-	return t.console.Close()
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if err := t.console.Close(); err != nil {
+		return err
+	}
+	return nil
```
Comment (phemmer) [inline daemon/execdriver/native/driver.go]: This can be shortened to just:

    return t.console.Close()

Comment (runcom) [inline daemon/execdriver/native/driver.go]: Done

**Comment by runcom** [comment]
`DockerSuite.TestRunNonRootUserResolvName` failing three times on CI but passes on local, any clue?

**Comment by LK4D4** [comment]
@runcom Could you check if there are no leaked goroutines after -it containers stop?

**Comment by runcom** [comment]
Checked before submitting this pr and IIRC at least one go routine was leaked :/

**Comment by runcom** [comment]
@LK4D4 this patch doesn't solve anything about goroutine leakage tho..
after -ti command there are 4 more FD and 4 goroutine leaked

**Comment by cpuguy83** [comment]
Hmm, this looks doesn't look like it would work.
`io.Copy` blocks until EOF, EOF won't happen because `Close` is also surrounded by lock... unless somewhere pipes.Stdout is getting closed.

**Comment by runcom** [comment]
oww I see

**Comment by LK4D4** [comment]
At least we know that we need to write some test :)

**Comment by runcom** [comment]
@cpuguy83 @LK4D4  I suppose, somehow, when pipes.stdout is closed with ctrl^d the write to stdout in https://golang.org/src/io/io.go#L364 errors out and break releasing the lock in io.Copy and Close gets executed....does this make sense?

**Comment by LK4D4** [comment]
@runcom I more concerned about application side of closing.

**Pull request closed without merging**